torch/testing/_internal/distributed/nn/api/remote_module_test.py

```diff
@@ -0,0 +1,130 @@
+import enum
+from typing import Tuple
+
+import torch
+import torch.testing._internal.dist_utils as dist_utils
+from torch import Tensor, nn
+from torch.distributed.nn.api.remote_module import RemoteModule, _gen_global_unique_name
+from torch.testing._internal.distributed.rpc.rpc_agent_test_fixture import (
+    RpcAgentTestFixture,
+)
+
+
+class ModuleCreationMode(enum.Enum):
+    SCRIPT_MODULE_CREATOR = "script_module_creator"
+    MODULE_CTOR_WITH_INTERFACE = "module_ctor_with_interface"
+    MODULE_CTOR = "module_ctor"
+
+
+@torch.jit.interface
+class MyModuleInterface:
+    def forward(
+        self, tensor: Tensor, number: int, word: str = "default"
+    ) -> Tuple[str, int, Tensor]:
+        pass
+
+
+class MyModule(nn.Module):
+    def __init__(self, first_arg, first_kwarg=-1):
+        super().__init__()
+
+    def forward(
+        self, tensor: Tensor, number: int, word: str = "default"
+    ) -> Tuple[str, int, Tensor]:
+        return word, number, tensor
+
+
+def create_module(first_arg, first_kwarg=-1) -> MyModuleInterface:
+    return MyModule(first_arg, first_kwarg=first_kwarg)
+
+
+class RemoteModuleTest(RpcAgentTestFixture):
+    @property
+    def world_size(self):  # Override setting in RpcAgentTestFixture
+        return 2
+
+    def setUp(self):
+        super().setUp()
+        self._fork_processes()
+
+    @dist_utils.dist_init
+    def test_gen_global_unique_name(self):
+        if self.rank != 0:
+            return
+        name_0 = _gen_global_unique_name()
+        name_1 = _gen_global_unique_name()
+        self.assertNotEqual(name_0, name_1)
+
+    @staticmethod
+    def _create_remote_module_iter(dst_worker_name, modes=None):
+        if modes is None:
+            modes = ModuleCreationMode.__members__.values()
+
+        args = (1,)
+        kwargs = dict(first_kwarg=2)
+        if ModuleCreationMode.SCRIPT_MODULE_CREATOR in modes:
+            remote_module = RemoteModule(dst_worker_name, create_module, args, kwargs)
+            scripted_remote_module = torch.jit.script(remote_module)
+            yield scripted_remote_module
+
+        if ModuleCreationMode.MODULE_CTOR_WITH_INTERFACE in modes:
+            remote_module = RemoteModule(
+                dst_worker_name,
+                MyModule,
+                args,
+                kwargs,
+                module_interface_cls=MyModuleInterface,
```

Comment:
Why do we need a module creator and a module interface. What if these two do not match?

Reply:
I feel we the user shouldn't have to pass in the ModuleInterface or a globally unique name.

Reply:
globally unique name is necessary for profiling purpose. We want to name all RemoteModules, so that we don't have to have a NamedEmbeddingBag wrapper and another NamedEmbedding wrapper. They can all use the RemoteModule, which takes a name.

Reply:
> Why do we need a module creator and a module interface. What if these two do not match?

Removing inferring module_interface_cls from module_creator to aovid confusion. Users always need to explicitly provide module_interface_cls if they want the RemoteModule to be torchscriptable.